### ISSUE TITLE
Shape inference issues after running dynamic_rnn

### ISSUE THREAD

**Issue by shtoshni** [tensorflow/tensorflow#7595]
I am trying to use dynamic_rnn in the encoder part of seq2seq library. I am using a placeholder for encoder input with dimensions [None, None] to reflect dynamic batch size and time steps. 

The problem is that the output that I get after the dynamic_rnn run has unknown shape along time dimension which causes issues when I try to use it along with attention. I have been able to overcome it by using the bucket length to set_shape but now that I am trying to ditch bucketing, it seems a little non-trivial. I was trying to use the maximum of sequence lengths but I have not been able to successfully use the value obtained from tf.reduce_max(sequence_length). Any help/suggestions would be much appreciated. Also, I am using version 0.11.

**Comment by yaroslavvb** [comment]
cc @ebrevdo who has substantial experience with it (he'll probably ask you to upgrade to TF 1.0 first)

**Comment by ebrevdo** [comment]
Sounds like your encoder inputs have unknown batch_size and time_steps,
right?  But you expect dynamic_rnn to "fill in" the unknown time_steps?
Since time_steps are not known at graph build time, you cannot fill it into
the static shape, even if you use reduce_max.  the result of reduce_max is
a Tensor, which you cannot use to set static shape.

It sounds like the attention mechanism wants a static number of time steps,
and is unable to handle variable length outputs.  That or you're somehow
not using the attention correctly.

That said, we hope to have a better decoder + attention mechanism coming
soon (next week or two); in tf.contrib.seq2seq, on tensorflow's github
master branch.

**Comment by shtoshni** [comment]
Thanks for such a quick response !
You are right that the dynamic_rnn can't fill in the unknown time steps but really it should stop at max of seq_len and can thus, fill in this dimension (am I expecting too much ?).
Yeah the attention mechanism currently requires the size of input to be known. Thanks for the heads up about the new version. But how are you guys planning on handling this. Is it a matter of upgrading Tensorflow or is there something that can be done in the current version itself ?

**Comment by ebrevdo** [comment]
Yup you're expecting too much :)  the dynamic_rnn will stop at the max of
seq_len, but that max can change from one minibatch to the next!  So it's
also not known.

The new attentional mechanism should be able to handle an unknown sequence
length; we'll keep this in mind as we're designing it.

On Thu, Feb 16, 2017 at 3:47 PM, Shubham Toshniwal <notifications@github.com

**Comment by shtoshni** [comment]
As it turns out, it was just a matter of avoiding certain reshape operations. So I just replaced those by expand_dim operations and things are working fine now. 

Using dynamic_rnn in encoder and raw_rnn in decoder, I was able to avoid the bucketing shenanigan. Again, thanks for the help and I will close this issue.

**Comment by ebrevdo** [comment]
Got it.  For posterity, can you tell us the original reshape operations and
the new expand_dim versions?  If you can expand_dim and preserve shape,
then presumably the reshape should work too.  This may be a shape inference
bug on our end.

**Comment by shtoshni** [comment]
I will just describe the parts I modified and the code snippet for that piece.

- attention_states has shape (batch_size, attn_length, attn_size) where attn_length is enc_inp_len
The first 2 dimensions are unknown and that will cause issue with reshaping ops trying to preserve the first 2 dimensions. However, the reshaping operation used in seq2seq below was easily replaceable by the next line. 
`hidden = array_ops.reshape( attention_states, [-1, attn_length, 1, attn_size])` 
`hidden = tf.expand_dims(attention_states, 2)`

- Multiplying attention weights `alpha` with encoder hidden states, `hidden`, requires another reshape in the original code
alpha has shape (batch_size, attn_length) and is calculated via
`alpha = nn_ops.softmax(s)` 
The original code uses the following line to get context vector **c**
`c = math_ops.reduce_sum( 
                        array_ops.reshape(alpha, [-1, attn_length, 1, 1]) * hidden,                                       
                        [1, 2])
`
I replaced it via
`alpha = tf.expand_dims(alpha, 2)` 
`alpha = tf.expand_dims(alpha, 3)` 
`c = math_ops.reduce_sum(alpha * hidden, [1, 2])`

Hope this helps!

**Comment by shtoshni** [comment]
By the way, since you guys are looking at improving the seq2seq library, there is one mathematical inaccuracy in current implementation. 
- The attention mechanism currently considers the whole padded input. 
- Even though the padded input won't contribute to context vector due to **0** output vector corresponding to it, it does reduce the probability mass of the actual inputs as it is part of the `softmax` calculation.
- Ideally, there should be a mask applied to the `softmax` output and subsequent re-normalization.  Something along these lines:
`attn_mask = tf.sequence_mask(seq_len_inp, dtype=tf.float32)`
`alpha = nn_ops.softmax(s) * attn_mask`
`alpha = alpha / (tf.reduce_sum(alpha, reduction_indices=[1]) + 1e-12)`
- However, I must say that things work out fine even without it but it's needed to match the math described in research papers.

**Comment by ebrevdo** [comment]
Thanks!  Forwarded these two notes (one on slicing to andrew; the other to
a collaborator working on attention).
